### PR TITLE
fix(cloud-element-templates): keep template properties order

### LINKED ISSUE #838: Input mappings defined by templates not stable
__Describe the Bug__

As a user I define element template properties in a dedicated order. As it looks like that order is not uphelt as input mappings are created. In fact it seems like we attach the last saved last in the `<zeebe:ioMapping>` list. This leads to dependent mappings not being evaluated consistently (https://github.com/camunda/web-modeler/issues/3251). 

capture NxgZqR_optimized

The whole behavior is bogus to the user, and cannot worked around without knowing.

__Steps to Reproduce__

1. configure [`dependent-io.json` template](https://github.com/bpmn-io/bpmn-js-properties-panel/files/10234754/dependent-io.json.txt)
2. Model `Dependent Mapping` task
2. Change values of `First*`
3. See that we move each one of them (`zebee:input`, `zeebe:output`, `zeebe:property`) below dependent property

__Expected Behavior__

* Upon modification we uphold the order of properties defined in the template.
* Upon entry addition (i.e. conditionally enabled) we uphold the order of properties defined in template.

__Environment__

- Host (Browser/Node version), if applicable: Any
- OS: Any
- Library version: 1.12.0

### PR DESCRIPTION
**Pull request by smbea**
Closes #838

We could consider refactoring `propertyUtil` later on to not delete the modified properties and just update the values instead, see https://github.com/bpmn-io/bpmn-js-properties-panel/issues/838#issuecomment-1369792405

### CODE REVIEW

**Review by marstamm — CHANGES_REQUESTED** [review]
Good work so far, the implementation works fine for the common use cases with the mappings already present in the XML. There are some edge cases where we have no explicitly defined behavior yet. Specifically, where should new entries be added?

With this PR (and https://github.com/bpmn-io/bpmn-js-properties-panel/pull/843) we are introducing the concept that "order matters". I think it is important to be clear about our story around it. My assumption is, that the order in the XML corresponds to what the order user sees in the properties panel. Is this assumption correct?
In that case, we should insert new and optional properties at the position according to the element template.

Does this make sense? Let me know what you think about this proposal.

File: src/provider/cloud-element-templates/util/propertyUtil.js [hunk 1]
```diff
@@ -360,12 +366,16 @@ export function setPropertyValue(bpmnFactory, commandStack, element, property, v
...
+      const index = taskHeaders.get('values').indexOf(oldTaskHeader);
+      values.splice(index, 0, newTaskHeader);
+
       commands.push({
         cmd: 'element.updateModdleProperties',
         context: {
           ...context,
           moddleElement: taskHeaders,
-          properties: { values: [ ...values, newTaskHeader ] }
+          properties: { values }
```
[the remark this answers is not in the document]
Comment (marstamm) [reply src/provider/cloud-element-templates/util/propertyUtil.js:378]: > We could consider refactoring this later on to not delete the properties here and just update the values instead, see https://github.com/bpmn-io/bpmn-js-properties-panel/issues/838#issuecomment-1369792405

I did not see your comment about this one, sorry 🙈

**Comment by nikku** [comment]
Let's move specs for this aspect to a dedicated document: https://github.com/bpmn-io/bpmn-js-properties-panel/compare/838-stable-property-order.

The tests I've added need to be expanded:

* Fix/reorder mappings on modification (automatically heal, if existing ones are broken)?
* New entries are added in the correct places (i.e. when non-existing initially _or_ contitionally activated)

File: test/spec/provider/cloud-element-templates/fixtures/complex.json [hunk 2]
```diff
@@ -683,5 +683,139 @@
         }
       }
     ]
+  },
+  {
```
Comment (nikku) [inline test/spec/provider/cloud-element-templates/fixtures/complex.json:687]: Let's not mangle it into `complex.json`, but instead keep it in a separate file, under a clear name, i.e. `update-properties-order.json`.

You can rely on all templates in `cloud-element-templates/fixtures` to be loaded by the test, cf. https://github.com/bpmn-io/bpmn-js-properties-panel/blob/master/test/spec/BpmnPropertiesPanelRenderer.spec.js#L245.

**Comment by nikku** [comment]
I'm end-to-end testing this today. Great work on finding a robust, self-healing solution @smbea @marstamm :medal_sports:.

**Review by nikku — COMMENTED** [review]
@marstamm You want to give your final approal on this? Looks good to me.

**Review by marstamm — APPROVED** [review]
Great work @smbea , works like a charm

**Pull request merged**